nexus/reconfigurator/simulation/src/system.rs

```diff
@@ -0,0 +1,473 @@
+                time_created: Utc::now(),
+                zones: vec![internal_dns],
+            }),
+        );
+        self.system.external_dns.insert(
+            blueprint.external_dns_version,
+            Arc::new(DnsConfigParams {
+                generation: blueprint.external_dns_version,
+                // TODO: probably want to make time controllable by the caller.
+                time_created: Utc::now(),
+                zones: vec![external_dns],
+            }),
+        );
+        self.system.blueprints.insert(
+            example.initial_blueprint.id,
+            Arc::new(example.initial_blueprint),
+        );
+        self.system.blueprints.insert(blueprint.id, Arc::new(blueprint));
+    }
+
+    pub fn add_collection(
+        &mut self,
+        collection: impl Into<Arc<Collection>>,
+    ) -> Result<(), DuplicateError> {
+        let collection = collection.into();
+        self.add_collection_inner(collection)
+    }
+
+    fn add_collection_inner(
+        &mut self,
+        collection: Arc<Collection>,
+    ) -> Result<(), DuplicateError> {
+        let collection_id = collection.id;
+        match self.system.collections.entry(collection_id) {
+            indexmap::map::Entry::Vacant(entry) => {
+                entry.insert(collection);
+                self.log.push(SimSystemLogEntry::AddCollection(collection_id));
+                Ok(())
+            }
+            indexmap::map::Entry::Occupied(_) => {
+                Err(DuplicateError::collection(collection_id))
+            }
+        }
+    }
+
+    pub fn add_blueprint(
+        &mut self,
+        blueprint: impl Into<Arc<Blueprint>>,
+    ) -> Result<(), DuplicateError> {
+        let blueprint = blueprint.into();
+        self.add_blueprint_inner(blueprint)
+    }
+
+    fn add_blueprint_inner(
+        &mut self,
+        blueprint: Arc<Blueprint>,
+    ) -> Result<(), DuplicateError> {
+        let blueprint_id = blueprint.id;
+        match self.system.blueprints.entry(blueprint_id) {
+            indexmap::map::Entry::Vacant(entry) => {
+                entry.insert(blueprint);
+                self.log.push(SimSystemLogEntry::AddBlueprint(blueprint_id));
+                Ok(())
+            }
+            indexmap::map::Entry::Occupied(_) => {
+                Err(DuplicateError::blueprint(blueprint_id))
+            }
+        }
+    }
+
+    pub fn add_internal_dns(
+        &mut self,
+        params: impl Into<Arc<DnsConfigParams>>,
+    ) -> Result<(), DuplicateError> {
+        let params = params.into();
+        self.add_internal_dns_inner(params)
+    }
+
+    fn add_internal_dns_inner(
+        &mut self,
+        params: Arc<DnsConfigParams>,
+    ) -> Result<(), DuplicateError> {
+        let generation = params.generation;
+        match self.system.internal_dns.entry(generation) {
+            std::collections::btree_map::Entry::Vacant(entry) => {
+                entry.insert(params);
```

Comment:
Should the internal/external DNS operations get entries in `self.log`?

Reply:
Ah so there's a distinction between internal and external callers here -- external callers add to logs while internal callers don't. But I realized the code wasn't clear about this at all. I've refactored it into having two layers: the external public API layer which does logging, and the internal layer which does not.